pkg/agent/multicluster/mc_route_controller.go

```diff
@@ -0,0 +1,556 @@
+	if err = c.addMCNodeRoutes(gw, ciImp, iAmActiveGW); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *MCRouteController) addMCNodeRoutes(gw *mcv1alpha1.Gateway, ciImp *mcv1alpha1.ClusterInfoImport, iAmActiveGW bool) error {
+	allciImports, err := getAllCIImports(ciImp, c.ciImportLister, c.namespace)
+	if allciImports == nil {
+		return err
+	}
+	var tunnelPeerIP net.IP
+	for _, ciImport := range allciImports {
+		tunnelPeerIP = getPeerGatewayIP(ciImport.Spec)
+		if tunnelPeerIP == nil {
+			klog.V(2).InfoS("No valid peer Gateway IP from ClusterInfoImport, skip updating openflow rules",
+				"clusterinfoimport", klog.KObj(ciImport), "node", c.nodeConfig.Name)
+			continue
+		}
+
+		klog.InfoS("Adding/Updating remote Gateway Node routes for Multi-cluster", "gateway", klog.KObj(gw),
+			"node", c.nodeConfig.Name, "peer", tunnelPeerIP)
+		allCIDRs := []string{ciImport.Spec.ServiceCIDR}
+		peerConfigs, err := generatePeerConfigs(allCIDRs, tunnelPeerIP)
+		if err != nil {
+			klog.ErrorS(err, "Parse error for serviceCIDR from remote cluster", "clusterinfoimport", ciImport.Name, "gateway", gw.Name)
+			return err
+		}
+		if iAmActiveGW {
+			klog.InfoS("Adding/Updating flows to remote Gateway Node for Multi-cluster traffic", "clusterinfoimport", ciImport.Name, "cidrs", allCIDRs)
+			localGatewayIP := net.ParseIP(gw.GatewayIP)
+			if err := c.ofClient.InstallMulticlusterGatewayFlows(
+				getCacheKeyWithPrefix(ciImport.Spec.ClusterID),
+				peerConfigs,
+				tunnelPeerIP,
+				localGatewayIP); err != nil {
+				return fmt.Errorf("failed to install flows to remote Gateway in ClusterInfoImport %s: %v", ciImport.Name, err)
+			}
+		} else {
+			tunnelPeerIP := net.ParseIP(gw.InternalIP)
+			if err := c.ofClient.InstallMulticlusterNodeFlows(
+				getCacheKeyWithPrefix(gw.Name+ciImport.Spec.ClusterID),
+				peerConfigs,
+				tunnelPeerIP); err != nil {
+				return fmt.Errorf("failed to install flows to Gateway %s: %v", gw.Name, err)
+			}
+		}
+
+		c.installedCIImport.Add(&ciImportInfo{
+			localGWName: gw.Name,
+			ciImport:    *ciImport,
+		})
+	}
+	c.installedGWName = gw.Name
+	return nil
+}
+
+// checkActiveGateway will compare Gateway's CreationTimestamp to get active Gateway,
+// and also check if given name is a Gateway. The last created Gateway will be the active Gateway.
+func checkActiveGateway(gwLister mclisters.GatewayLister, namespace string, myName string) (*mcv1alpha1.Gateway, bool, error) {
+	gwlist, err := gwLister.Gateways(namespace).List(labels.Everything())
+	if err != nil {
+		klog.ErrorS(err, "Failed to get Gateway list", "namespace", namespace)
+		return nil, false, err
+	}
+	if len(gwlist) == 0 {
+		return nil, false, fmt.Errorf("no Gateway found in Namespace %s", namespace)
+	}
+	var iAmGW bool
+	for _, gw := range gwlist {
+		if gw.Name == myName {
+			iAmGW = true
+			break
+		}
+	}
+	// Sort Gateways by CreationTimestamp. The last created Gateway will be the first element.
+	sort.Slice(gwlist, func(i, j int) bool {
```

Comment:
> It is more efficient to iterate the slice and get the last created one.

I meant instead of sorting the slice, you can use a loop to iterate all Gateways in the slice and return the the last created one, and so it is O(1) and no extra memory consumption.

Reply:
Done